### PR TITLE
Debugoutput

### PR DESCRIPTION
**Pull request by sbird**
This PR saves a variety of extra fields (all fields in P and SPHP) to the crash dump snapshot, and optionally to the main snapshot. It also saves and restores DelayTime, which we weren't doing before. This should be enough to debug future "bad timestep" crashes.

### CODE REVIEW

File: libgadget/petaio.c [hunk 1]
```diff
@@ -856,10 +859,47 @@ static void register_io_blocks() {
...
+    IO_REG_WRONLY(DtEntropy,       "f4", 1, 0);
+    IO_REG_WRONLY(DhsmlEgyDensityFactor,       "f4", 1, 0);
+    IO_REG_WRONLY(DivVel,       "f4", 1, 0);
+    IO_REG_WRONLY(CurlVel,       "f4", 1, 0);
     /*Sort IO blocks so similar types are together; then ordered by the sequence they are declared. */
-    qsort_openmp(IOTable.ent, IOTable.used, sizeof(struct IOTableEntry), order_by_type);
+    qsort_openmp(IOTable->ent, IOTable->used, sizeof(struct IOTableEntry), order_by_type);
+}
+
+void free_io_blocks(struct IOTable * IOTable) {
```
Comment (rainwoodman) [inline libgadget/petaio.c]: destroy_io_table(); or perhaps purge_io_table().

destroy: frees the members.
free: frees the object and the member.
purge: nothing to do with life cycle of the object.

File: libgadget/petaio.h [hunk 2]
```diff
@@ -55,20 +67,21 @@ petaio_build_selection(int * selection,
  * IO_REG_WRONLY declares an io block which is written, but is not read on snapshot load.
  * */
 #define IO_REG(name, dtype, items, ptype) \
```
Comment (rainwoodman) [inline libgadget/petaio.h]: Perhaps worth it to plumb IOTable up through these macros -- a bit of search and replacement, but avoids an hidden argument.

File: libgadget/petaio.c (hunk 1 above)
Comment (sbird) [reply libgadget/petaio.c]: Done!

File: libgadget/petaio.h (hunk 2 above)
Comment (sbird) [reply libgadget/petaio.h]: Yes, but it is a fair bit of careful search and replace. I don't really want to just now.

**Comment by sbird** [comment]
@yueyingn : Can you re-run the crashing simulation with this PR included and find the low-timebin particles in the output?

File: libgadget/petaio.h (hunk 2 above)
Comment (sbird) [reply libgadget/petaio.h]: I did it anyway :-)

**Pull request merged**